### PR TITLE
ARROW-15860: [Python] Document RecordBatchReader

### PR DESCRIPTION
**Pull request by wjones127**

### CODE REVIEW

File: docs/source/python/api/tables.rst [hunk 1]
```diff
@@ -44,6 +44,7 @@ Classes
    RecordBatch
    Table
    TableGroupBy
+   lib.RecordBatchReader
```
Comment (amol-) [inline docs/source/python/api/tables.rst:47]: We shouldn't document things from the `lib` namespace, mostly those are internals.
Also given that the `RecordBatchReader` comes from the ipc implementation module, it should probably get exposed through `pyarrow.ipc` like all the other readers.

File: python/pyarrow/_dataset.pyx [hunk 2]
```diff
@@ -2357,7 +2357,12 @@ cdef class Scanner(_Weakrefable):
         return GetResultValue(result)
 
     def to_reader(self):
-        """Consume this scanner as a RecordBatchReader."""
+        """Consume this scanner as a RecordBatchReader.
+        
+        Returns
+        -------
+        pyarrow.lib.RecordBatchReader
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx:2364]: Also here use the public module (so `pyarrow.lib.RecordBatchReader` I suppose)

**Review by jorisvandenbossche — COMMENTED** [review]
I think it would be good to update the docstring of `RecordBatchReader` a bit, if we are going to include it in the online docs (well, updating it is useful anyway). 
Currently, it mentions "Base class for reading stream of record batches ...  Should not
be instantiated directly by user code". But, it seems we nowadays have a `from_batches` method to actually construct a RecordBatchReader in user code? It might also be useful to give an example where this is being returned (I think nowadays it is no longer purely a base class for `RecordBatchStream/FileReader`, but actually also gets returned in some cases?)

As @amol- mentions, we should indeed document it how it is exposed publicly, so `pyarrow.ipc`. But isn't that actually a bit a strange place to expose this? As purely the RecordBatchReader has not really anything to do with IPC? (its subclasses `RecordBatchStream/FileReader` do, but now we use the base class directly as well, it might make sense exposing this in the top-level namespace instead?)

**Comment by wjones127** [comment]
> I think it would be good to update the docstring of RecordBatchReader a bit, if we are going to include it in the online docs (well, updating it is useful anyway).

Yeah, I likely need to provide an example of iterating over it. Anything else you  want to see.

> we should indeed document it how it is exposed publicly, so pyarrow.ipc. But isn't that actually a bit a strange place to expose this? As purely the RecordBatchReader has not really anything to do with IPC? (its subclasses RecordBatchStream/FileReader do, but now we use the base class directly as well, it might make sense exposing this in the top-level namespace instead?)

Yes my understanding is that RBR reader is a type that will be returned by a variety of functions in PyArrow. It's currently returned from `Dataset.Scanner.to_reader()`, various flight readers, and some IPC methods. There are also functions like `write_dataset()` that take RBR as arguments. 

I wanted to document it so other libraries could know how to consume them.

Would it be possible to move it out of the lib module?

**Comment by wjones127** [comment]
> It's actually not needed to move it out of the lib module, but it's rather a matter of in which public python file it gets imported to expose it publicly (there are also many objects from lib that are imported in the main __init__.py file)

Sorry that's what I meant. To import it so it's "moved" in the API, so we aren't importing it from lib.

IMO It belongs in `__init__.py` next to RecordBatch. It's used across the different submodules, and in the C++ implementation it's in the same namespace and file as RecordBatch.

**Comment by jorisvandenbossche** [comment]
> IMO It belongs in `__init__.py` next to RecordBatch.

Agreed. 

I think it is fine to move it to the top-level `__init__.py`. We can keep it in `pyarrow/ipc.py` for a while for backwards compatibility, or even explicitly deprecate accessing it from that namespace.

File: python/pyarrow/ipc.pxi [hunk 3]
```diff
@@ -550,8 +550,36 @@ class _ReadPandasMixin:
...
+    To import and export using the Arrow C stream interface, use the 
+    ``_import_from_c`` and ``_export_from_c`` methods. However, keep in mind this
+    interface is experimental and intended for expert users.
+
+    Examples
+    --------
+    >>> schema = pa.schema([('x', pa.int64())])
+    >>> def iter_record_batches():
+    >>>     for i in range(2):
+    >>>     yield pa.RecordBatch.from_arrays([pa.array([1, 2, 3])], schema=schema)
```
Comment (jorisvandenbossche) [inline python/pyarrow/ipc.pxi:572]:
```suggestion
    >>> def iter_record_batches():
    ...     for i in range(2):
    ...     yield pa.RecordBatch.from_arrays([pa.array([1, 2, 3])], schema=schema)
```

File: python/pyarrow/ipc.pxi [hunk 4]
```diff
@@ -550,8 +550,36 @@ class _ReadPandasMixin:
...
+    >>> reader = pa.RecordBatchReader.from_batches(schema, iter_record_batches())
+    >>> print(reader.schema)
+    pyarrow.Schema
+    x: int64
+    >>> for batch in reader:
+    >>>     print(batch)
```
Comment (jorisvandenbossche) [inline python/pyarrow/ipc.pxi:578]:
```suggestion
    >>> for batch in reader:
    ...     print(batch)
```

**Review by jorisvandenbossche — COMMENTED** [review]
Nice doc updates!

Can you also update the tests to use the new import (we should at least test that somewhere), and maybe keep one test ensuring we can still import it from pyarrow.ipc as well for back compat)

**Comment by wjones127** [comment]
> Can you also update the tests to use the new import (we should at least test that somewhere), and maybe keep one test ensuring we can still import it from pyarrow.ipc as well for back compat)

Sure! I changed for all tests except the IPC ones for now.

File: docs/source/python/data.rst [hunk 5]
```diff
@@ -512,3 +512,23 @@ a new schema and cast the data to this schema:
...
 and so they are bytes objects (``b'...'``) in Python.
+
+Record Batch Readers
+--------------------
+
+Many functions in PyArrow either return or take as an argument a :class:`RecordBatchReader`.
+It can be used like any iterable of record batches, but also provides their common
+schema without having to get any of the batches.
+
+.. ipython:: python
```
[the remark this answers is not in the document]
Comment (pitrou) [reply docs/source/python/data.rst:523]: If not doctest syntax (does it need an extension to be enabled and/or some scaffolding to be set up?), then a plain Python code block would do. I wouldn't want to accumulate more technical debt.

[the remark this answers is not in the document]
Comment (wjones127) [reply docs/source/python/data.rst:523]: I see doctest extension is included, but I see no usage of doctest directive yet.

I agree with Antoine that I don't want this contribute to tech debt, but it seems if I use a plain code block or ipython either way we will want to convert to a doctest in the future. So inclined to leave as is, but happy to help make a PR soon for moving to doctest.

[the remark this answers is not in the document]
Comment (wjones127) [reply docs/source/python/data.rst:523]: Oh okay. I'll transition that then. Thanks @jorisvandenbossche

File: docs/source/python/data.rst [hunk 6]
```diff
@@ -518,17 +518,15 @@ Record Batch Readers
...
-   for batch in reader:
-      print(batch)
+schema without having to get any of the batches.::
+
+   >>> schema = pa.schema([('x', pa.int64())])
+   >>> def iter_record_batches():
+   ...    for i in range(2):
+   ...       yield pa.RecordBatch.from_arrays([pa.array([1, 2, 3])], schema=schema)
+   >>> reader = pa.RecordBatchReader.from_batches(schema, iter_record_batches())
+   >>> print(reader.schema)
```
Comment (jorisvandenbossche) [inline docs/source/python/data.rst:528]: You can add the output here of the print statement (so users see what the code does)

File: docs/source/python/data.rst [hunk 7]
```diff
@@ -518,17 +518,15 @@ Record Batch Readers
...
+   >>> for batch in reader:
+   ...    print(batch)
```
Comment (jorisvandenbossche) [inline docs/source/python/data.rst:530]: Same here

File: docs/source/python/data.rst (hunk 6 above)
Comment (wjones127) [reply docs/source/python/data.rst:528]: Thanks! Sorry I forgot that.

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request closed without merging**